climada/hazard/tc_tracks.py

```diff
@@ -1237,17 +1241,16 @@ def to_geodataframe(self, as_points=False):
         if as_points:
             gdf_long = pd.concat([track.to_dataframe().assign(idx=i)
                                   for i, track in enumerate(self.data)])
-            gdf_long['geometry'] = gdf_long.apply(lambda x: Point(x['lon'],x['lat']), axis=1)
+            gdf_long['lon'] = u_coord.lon_normalize(gdf_long['lon'].values.copy())
+            gdf_long['geometry'] = gdf_long.apply(lambda x: Point(x['lon'], x['lat']), axis=1)
             gdf_long = gdf_long.drop(columns=['lon', 'lat'])
             gdf_long = gpd.GeoDataFrame(gdf_long.reset_index().set_index('idx'),
                                         geometry='geometry')
             gdf = gdf_long.join(gdf)
 
         else:
-            # LineString only works with more than one lat/lon pair
             gdf.geometry = gpd.GeoSeries([
-                LineString(np.c_[track.lon, track.lat]) if track.lon.size > 1
-                else Point(track.lon.data, track.lat.data)
+                _make_line_from_track(track, split_lines_antimeridian=split_lines_antimeridian)
                 for track in self.data
             ])
```

Comment:
I suggest to remove `_make_line_from_track` and use `shapely.ops.split` instead:
```
            gdf.geometry = gpd.GeoSeries([
                LineString(np.c_[track.lon, track.lat]) if track.lon.size > 1
                else Point(track.lon.data, track.lat.data)
                for track in self.data
            ])
            gdf.crs = DEF_CRS
            if split_lines_antimeridian:
                t_mask = np.asarray([
                    (t.lon > 170).any() and (t.lon < -170).any() and t.lon.size > 1
                    for t in self.data])
                gdf.loc[t_mask, "geometry"] = gdf.geometry[t_mask] \
                    .to_crs({"proj": "longlat", "lon_wrap": 180}) \
                    .apply(lambda line: shapely.geometry.MultiLineString([
                        shapely.geometry.LineString([(x - 360, y) for x, y in segment.coords])
                        if any(x > 180 for x, y in segment.coords) else segment
                        for segment in shapely.ops.split(line, antimeridian).geoms
                    ]))
```

Reply:
The problem with the shapely ops is that you introduce unmergeable `MultiLineString`s on self-intersections of the underlying LineString - and since performance-wise, the difference is only small (@bguillod's version even wins out in my testing), I would propose to leave it as is and fix the `_make_line_from_track`.

Reply:
I don't understand the problem with self-intersections. Would you please add a unit test for that.

Reply:
When a track intersects itself, the shapely operation will cut the `LineString` into a `MultiLineString` at the points of intersection. Benoit's homebrewed function avoids that and _only_ introduces breaks into the original line where it crosses the antimeridian. The issue is documented [here](https://github.com/Toblerity/Shapely/issues/572). The `shapely.ops.linemerge` does not help to stitch/meld the lines back together again.

In this concrete case an example looks like this:
![image](https://user-images.githubusercontent.com/24587121/113105632-1384eb00-9202-11eb-9725-b6eef6614454.png)

And the `MultiLineString` coordinates around that track section that your code produces:

```
…
  (152.3000030517578, -23.200000762939453),
  (152.47250366210938, -23.279993057250977),
  (152.5104058915784, -23.285940228219452)],
 [(152.5104058915784, -23.285940228219452),
  (152.5702812971104, -23.29533517090472)],
 [(152.5702812971104, -23.29533517090472),
  (152.60000610351562, -23.299999237060547),
  (152.60571764022595, -23.292983514697355)],
 [(152.60571764022595, -23.292983514697355),
  (152.6875457763672, -23.19247055053711),
  (152.6999969482422, -23.100000381469727),
  (152.58509826660156, -23.185026168823242),
  (152.5104058915784, -23.285940228219452)],
 [(152.5104058915784, -23.285940228219452),
  (152.5, -23.299999237060547),
  (152.5702812971104, -23.29533517090472)],
 [(152.5702812971104, -23.29533517090472),
  (152.60571764022595, -23.292983514697355)],
 [(152.60571764022595, -23.292983514697355),
  (152.61257934570312, -23.29252815246582),
  (152.8000030517578, -23.299999237060547),
  …
  ```

Reply:
I will write a unit test, let me see if we already have some tracks from that region in the repo.

Reply:
I see that this seems unexpected, but at the same time, it's not that bad: The docs to the shapely.ops.split function don't claim to preserve line strings. It just says that geometries are split at the desired locations and a `GeometryCollection` is returned. That is done successfully, also in your example. For plotting purposes and intersections with other geometries that will work just fine. I can't even think about a real usecase where this will be a problem. We could even use this workaround the shapely issue: https://github.com/Toblerity/Shapely/issues/1068#issuecomment-770296614

However, I understand that the result is unsatisfactory to you and would concede until the shapely issue is resolved. In this case, I insist that you add really rigorous testing. And I will probably have a few more review comments about the current implementation ...